include/std/functional.h

```diff
@@ -0,0 +1,119 @@
+/**
+ * avr-libcpp
+ *
+ * Copyright 2020-2021, Andrew Countryman <apcountryman@gmail.com> and the avr-libcpp
+ * contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
+ * file except in compliance with the License. You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+/**
+ * \file
+ * \brief functional interface.
+ */
+
+#ifndef STD_FUNCTIONAL_H
+#define STD_FUNCTIONAL_H
+
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+namespace std {
+
+namespace Reference_Wrapper_Implementation {
+
+template<typename T>
+void test( T & ) noexcept;
+
+template<typename T>
+void test( T && ) = delete;
+
+} // namespace Reference_Wrapper_Implementation
+
+template<typename T>
+class reference_wrapper {
+  public:
+    template<
+        typename U,
+        typename = decltype( Reference_Wrapper_Implementation::test<T>( declval<U>() ) ),
+        typename = enable_if_t<not is_same_v<decay_t<U>, reference_wrapper>>>
+    reference_wrapper( U && u ) noexcept(
+        noexcept( Reference_Wrapper_Implementation::test<T>( declval<U>() ) ) ) :
+        m_pointer{ pointer( std::forward<U>( u ) ) }
+    {
+    }
+
+    reference_wrapper( reference_wrapper const & original ) noexcept = default;
+
+    auto operator=( reference_wrapper const & expression ) noexcept -> reference_wrapper & = default;
+
+    operator T &() const noexcept
+    {
+        return *m_pointer;
+    }
+
+    auto get() const noexcept -> T &
+    {
+        return *m_pointer;
+    }
+
+    template<typename... Arguments>
+    auto operator()( Arguments &&... arguments ) const -> invoke_result_t<T &, Arguments...>
+    {
+        return ( *m_pointer )( forward<Arguments>( arguments )... );
+    }
+
+  private:
+    T * pointer( T & t ) noexcept
```

Comment:
This function should be a static member function and should use trailing return type syntax.